src/server/redis_connection.cc

```diff
@@ -498,9 +498,24 @@ void Connection::ExecuteCommands(std::deque<CommandTokens> *to_process_cmds) {
 
     SetLastCmd(cmd_name);
     {
+      std::optional<MultiLockGuard> guard;
```

Comment:
what raise my curiousity is that, would this be dangerous on blocking command? Should we set a "relock-callback" for them?

Reply:
Yeah we need to ensure it works well on blocking commands.

Reply:
Hmm it does need some work.

Like such a workflow:
```
lock
ExecuteCommand (block and return)
unlock
lock
Callback (timeout or satisfied)
unlock
```

Reply:
I wonder how callback "lock", previously, lock is implemented by internal cmd. Now, internal cmd doesn't lock, and we have "BlockKeys" rather than "lock-all-keys"?

Reply:
I think we can extract a "MutexBuilder" for blocking command 🤔 it could rebuild the mutex. But other solving is also ok to me

Reply:
Thanks for the reminder, the callback of the blocking command bypasses the `MultiLockGuard` in `ExecuteCommands`, and if a write operation is performed, we need to re-lock them at `OnWrite/OnBlockingWrite`.
```
  bool OnBlockingWrite() override {
    std::vector<std::string> lock_keys;
    lock_keys.reserve(keys_.size());
    for(const auto &key : keys_) {
      auto ns_key = AppendNamespace(key);
      lock_keys.emplace_back(std::move(ns_key));
    }
    MultiLockGuard guard(srv_->storage->GetLockManager(), lock_keys);
    engine::Context ctx(srv_->storage);
    auto s = TryPopFromList(ctx);
    return !s.IsNotFound();
  }
```  

At the same time, I think we need to be concerned not only about blocking commands, but also about other commands that use the `EventCallbackBase` mechanism, especially when it comes to `OnWrite`. `XReadGroup`, for example, looks like it's having a similar issue.

Also, could you talk more about `MutexBuilder` in detail? I'm not quite sure about the difference between what you think and what I think

Reply:
> Also, could you talk more about MutexBuilder in detail? I'm not quite sure about the difference between what you think and what I think

It can be simply pushing the keys into the blocking-command context, and when the execution resume, the keys can be resume. The point is:
1. The lock might be released to make other commands able to be executed.
2. The lock can be re-acquire (without cascade or ReentrantLock locking, like lock a outside then lock a internal)